### DISCUSSION TITLE
Support Swift for new Architecture

### DISCUSSION THREAD

**Discussion by maitrungduc1410** [callstack/react-native-builder-bob#425] [Ideas]
When using the CLI I don't see Swift for Turbo and Fabric options.

Not sure how can I try new architecture with Swift.

Thanks

**Comment by atlj** [comment]
Hey I was thinking about it for a while now. Since you cannot use codegen with Swift right now, there is a bit of overhead while writing Swift on the new architecture.
One solution would be to write your logic in swift, expose the methods through the @objc macro, and call them inside the codegen generated objc methods.
It is totally possible to create an objc project and add swift manually but I'll look into how we can integrate this directly into `create-react-native-library`.

**Reply by maitrungduc1410** [reply]
Hi, thanks for your response.

I got your details. I can still go with Swift and legacy arc for now.

Good work :)

**Comment by RobertSasak** [comment]
Here is my attempt. 

### Calculator.swift

Create a swift file with a simple sum instead of multiply. 

```swift
import Foundation

@objcMembers class Calculator: NSObject {

    func sum(numberOne: Double, numberTwo: Double) -> NSNumber {
        return (numberOne + numberTwo) as NSNumber
    }
    
}
```

### StripeApplePay.mm

Add if else Swift import in the main mm file. Notice how dash has changed to underscore followed by dash again. Initialize Swift class _Calculator_

```mm
#import "StripeApplePay.h"

// Added
#if __has_include(<react_native_stripe_apple_pay/react_native_stripe_apple_pay-Swift.h>)
#import <react_native_stripe_apple_pay/react_native_stripe_apple_pay-Swift.h>
#else
#import "react_native_stripe_apple_pay-Swift.h"
#endif

@implementation StripeApplePay
RCT_EXPORT_MODULE()

Calculator *calculator = [[Calculator alloc] init]; // Added

// Don't compile this code when we build for the old architecture.
#ifdef RCT_NEW_ARCH_ENABLED
- (NSNumber *)multiply:(double)a b:(double)b {
    // Calling Swift
    NSNumber *result = [calculator sumWithNumberOne:a numberTwo:b]; // Added
    // Calling C++
    //NSNumber *result = @(stripeapplepay::multiply(a, b));

    return result;
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params
{
    return std::make_shared<facebook::react::NativeStripeApplePaySpecJSI>(params);
}
#endif

@end

```


### react-native-stripe-apple-pay-Bridging-Header.h

Create emtpy bridge header file.

```h
// empty
```

Run your app and verify that 7+3 = 10 instead of 21.

@atlj Can I try to make a PR?

**Reply by atlj** [reply]
Hey @RobertSasak thanks a lot for the proposal! I've forgot to post here but we are actually working on this right now. You can find a really early work in this repo with the method you've posted https://github.com/atlj/react-native-kotlin-swift-backward-compatible-module/blob/dc6471bc52c7ae66846688c9bae68e126342434e/ios/KotlinSwiftBackwardCompatibleModule.mm. Please feel free to create a PR. I'm working on simplifying this approach even further by somehow getting rid of the mm file entirely. And I believe we can merge your approach for the moment being.

**Comment by maitrungduc1410** [comment]
somehow the Expo team does this seamlessly:

https://docs.expo.dev/guides/new-architecture/#my-build-failed-after-enabling-the-new

https://docs.expo.dev/modules/native-view-tutorial/

With Kotlin + Swift, as a developer I mostly don't see any diff in source code between old and new arch

**Comment by stanwolverine** [comment]
Any progress in this?

**Comment by maitrungduc1410** [comment]
it's 2025, it'll be great if we can have this 🙏
